### PR TITLE
Redo numpy.pxd changes and update tests

### PR DESCRIPTION
**Pull request by mattip**
xref gh-2498 (enables closing it?)
reverts gh-3114 which reverted gh-3095 (since the first version of the `@property` decorator was not complete)

Use the new `@property` decorator to support converting numpy python code like `ndarray.shape[0]` to `PyArray_DIMS(ndarray)[0]`. Now running `CFLAGS=-Wall python runtests.py numpy` produces no warnings. 

Also removed the unneeded `numpy_common.pxi`: numpy headers have apparently been cleaned up and no longer warn.

To make sure, I used this version via `pip install .` to build scipy which succeeded. Scipy still warns, once cython3.0 is released I can update the code in scipy to use `('NPY_NO_DEPRECATED_API', 'NPY_1_7_API_VERSION')`

### CODE REVIEW

**Comment by mattip** [comment]
All checks are passing

File: Cython/Includes/numpy/__init__.pxd [hunk 1]
```diff
@@ -244,14 +244,27 @@ cdef extern from "numpy/arrayobject.h":
...
-            # One should use PyArray_* macros instead to access the C fields.
             char *data
-            int ndim "nd"
-            npy_intp *shape "dimensions"
-            npy_intp *strides
             dtype descr  # deprecated since NumPy 1.7 !
             PyObject* base
 
+            @property
+            cdef int ndim(self):
```
Comment (scoder) [inline Cython/Includes/numpy/__init__.pxd]: Just wondering, but do you think it would help to declare these `inline`?

**Comment by scoder** [comment]
Lovely! :)

File: Cython/Includes/numpy/__init__.pxd (hunk 1 above)
Comment (mattip) [reply Cython/Includes/numpy/__init__.pxd]: maybe. I would have to look at the C-code from a real usage to see how much overhead would be removed

**Comment by mattip** [comment]
Thanks!

**Pull request merged**